cmd/api/src/api/signature.go

```diff
@@ -65,21 +68,81 @@ func tee(reader io.Reader, outA, outB io.Writer) error {
 	}
 }
 
-func GenerateRequestSignature(token, datetime, hmacMethod, requestMethod, requestURI string, body io.Reader) (io.Reader, []byte, error) {
-	if hmacMethod != auth.HMAC_SHA2_256 {
-		return nil, nil, fmt.Errorf("unsupported HMAC method: %s", hmacMethod)
+func NewSelfDestructingTempFile(dir, prefix string) (*SelfDestructingTempFile, error) {
+	if file, err := os.CreateTemp(dir, prefix); err != nil {
+		return nil, err
+	} else {
+		return &SelfDestructingTempFile{
+			file:      file,
+			destroyed: false,
+		}, err
+	}
+}
+
+type SelfDestructingTempFile struct {
+	file      *os.File
+	destroyed bool
+}
+
+func (s *SelfDestructingTempFile) SelfDestruct() error {
+	if !s.destroyed {
+		name := s.file.Name()
+
+		if err := s.file.Close(); err != nil {
+			return err
+		} else if err := os.Remove(name); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		s.destroyed = true
+	}
+	return nil
+}
+
+func (s *SelfDestructingTempFile) Read(b []byte) (int, error) {
+	n, err := s.file.Read(b)
+	if err == io.EOF {
+		s.SelfDestruct()
```

Comment:
This might have some weird side-effects if the file is read to `EOF` but `Seek` is then called. Thoughts?

Reply:
When the file is read to EOF or explicitly closed the file is intended to be closed and deleted. This way, if a payload is "large" and we have to stream the request body to disk for validation, we can clean up after ourselves without having to coordinate any additional steps with the subsequent middleware/handler beyond what it "should" do normally.

So, if the SelfDestructingTempFile is read to EOF the underlying file descriptor should be released and a subsequent call to Seek I would expect to fail in the same way as any other closed os.File.

Is there some side-effect I'm not seeing?